### DISCUSSION TITLE
Update icons for ID 5

### DISCUSSION THREAD

**Discussion by Vasmares** [BimmerGestalt/AAIdrive#428] [Ideas]
Hi,

have driven the 2016 F54 with ID4 and now I have the 2020 MINI F54 with ID5
After using your brillant app I came to notice, that some of the icons are still looking like ID4.

In the readme it says, that the Icons in the main menu cant be changed. 
![grafik](https://user-images.githubusercontent.com/10863304/136103513-abde0359-7c57-4775-a3f8-6fd381245161.png)

Fair enough, but what about the Icons in the Audio Player ?
Like the Search Icon, the "App List" Icon or the default cover art.
![grafik](https://user-images.githubusercontent.com/10863304/136103546-8f0ca013-8902-4c7f-ac25-37292e24eb45.png)

In the ID5 connect for iOS the design look more continous
![grafik](https://user-images.githubusercontent.com/10863304/136103731-cfd8ea5a-856f-492e-a917-bf6e6eba03fc.png)

Can the Icons be changed ?
If yes, did someone provide them ?
Is there like an "ID4 resource pack" ?

**Comment by hufman** [comment]
From the resources I've discovered, it is impossible to change those icons: they are forced to be the original icon pack provided by BMW.

The various image components around the car interface can be backed by two types of images. RaImageModel images can receive any image file from the phone, and this is found in the coverart component for example. The other type is an ImageIdModel, found in the toolbar icons and the artist/album label icons, which can only show images from an icon pack file that is uploaded at the start of the app connection. If you look within the APK file, you'll find an assets/carapplications directory with all of the resources that are sent to the car, including these icon packs.
The main menu icon (representing Audioplayer in the Media list, in this case), is such an ImageIdModel, and has the additional limitation that it can't even be changed to point to a different icon in the pack; it is hardcoded to a single specific icon. The main menu label is same idea, but with a TextIdModel.

The next restriction is that these resources are cryptographically locked to the app cert that is used to log in to the car. The multimedia app cert, for example, includes a sha256 checksum of the multimedia images.zip icon pack that it sends to the car, and so this images.zip file can not be altered.
There is one single point of flexibility: The Audioplayer app cert (and most others) allows both the BMW and the Mini icon packs, so it would be possible to use the other brand's icons (by changing [this line](https://github.com/BimmerGestalt/AAIdrive/blob/74f0998d94bfd0d26fa99e68285ba6c688e78e49/app/src/main/java/me/hufman/androidautoidrive/carapp/music/MusicApp.kt#L153) to use `"bmw"` instead of `iDriveConnectionStatus.brand`). But, the icon packs in BMW/Mini Connected are shared for all IDrive versions, so there isn't an ID4 vs ID5 icon pack in the resources that I've found.

Now, what you might be noticing is that your ID5 car is still using the ID4 Audioplayer layout, which looks a little small on your new bigger screen. There is indeed an upgraded ID5 multimedia layout, and AAIdrive uses the Spotify app cert to implement it. In AAIdrive's phone interface, it will indicate this functionality by showing `Global Coverart` as supported or not, with a reason for why it didn't decide to automatically switch to this new layout (and there's an advanced option to force it on). The Spotify icon pack includes the same ID4 toolbar icons, plus the Spotify main menu icon instead of the generic Audioplayer icon, and also some extras such as the shuffle and repeat toolbar icons.
The big improvement with this new AudioHmiState is that it matches the layout of the other music sources in the car, and so the Artist and Album icons are provided by the car itself, instead of using icons from the Audioplayer icon pack. Another difference is that the Song Title icon is (essentially) a RaImageModel and so we can fix the color format for that icon to match the rest of the interface.

I'm not sure the iOS picture made it through. I haven't used the iOS versions so I'd love to check it out! Let me know if I can explain anything else further!

**Reply by Vasmares** [reply]
Hi,
Ok I get you I will make some "iOS" ID5 Screenshots and attach them.

I updated Spotify today, and the layoput changed to what it used to look on iOS.
The icons still look like ID4.

I now know why I didnt update Spotify.

The sorting.

I had version 8.4.96.953 of spotify installed, (as you stated in the readme), because the current spotify app does not support the "custom sorting" layout. 
I sorts the playlists and folders alphabetical, and not like I have sorted them in the desktop app.
The old one did, but the global coverart wasnt working.

idk if spotify will fix this anytime soon....

**Reply by hufman** [reply]
Oh very curious!
Actually I should update the auto-detection logic to allow Global Coverart even with the older Spotify, as long as BMW/Mini Connected is not installed. Until then, you can use the advanced settings to Force Spotify Layout.

**Reply by Vasmares** [reply]
Would be very nice.
Does this mean the MINI Connected App cannot be installed ?
I have the MINI App and the AAiDrive App installed.

**Reply by hufman** [reply]
The reasoning is that AAIdrive shouldn't use the Spotify car layout (which is needed for Global Coverart) if the official Spotify car app is going to add its own duplicate icon in the car. However, since that feature was added, the BMW/Mini Connected apps have been shut down, which were required to run the official Spotify integration. I should update the feature autodetection to launch the new Spotify car layout even if old Spotify is installed as long as Connected is not installed too.

This won't change the requirement for installing Connected vs MyBMW/Mini! You could still install BMW/Mini Connected 6.4, and the Apps support will continue to work even without being able to log in to the app, if you want to experiment with the previous official apps. But otherwise, I think the new MyBMW/Mini apps work well enough to not need the Connected apps installed.

**Reply by hufman** [reply]
I pushed up a [change](https://github.com/BimmerGestalt/AAIdrive/commit/38887dbc925b4e9734125904e1a02c404aecf0db) that should (verified only with unit tests) enable the Spotify layout with the old Spotify installed if you don't have Connected installed, try it out!

**Comment by Vasmares** [comment]
Hi, 

sounds nice, Ill have installed it and will try as soo as I leave tomorrow morning.

I have attahed some pictures of what the UI looks like when an iPhone is connected:
![id5_ios_spotify_01](https://user-images.githubusercontent.com/10863304/136627496-cf6ed05e-7e61-4812-8610-96a77c3df2fe.jpg)
![id5_ios_spotify_02](https://user-images.githubusercontent.com/10863304/136627498-35ebc4c6-8275-456e-868d-af7a4f9ae2c0.jpg)
![id5_ios_audible_01](https://user-images.githubusercontent.com/10863304/136627499-320b856d-c8b7-4bcf-a812-98b43bc33766.jpg)

**Comment by Vasmares** [comment]
Regaringthe icons for ID5 i would like to make 2 suggestions.
There are 3 "better" icons that would fit nicely for ID5

See attached screenshot.

![id5_gui_new](https://user-images.githubusercontent.com/10863304/136628570-e46d44a1-4d13-4740-9029-7c00cbc1203f.png)

**Reply by hufman** [reply]
Thank you very much for the screenshots and the suggestions! I really appreciate the clear diagrams of your ideas!
Unfortunately, the car's resource system means that only the icons within the Spotify icon pack can be used, and it sadly doesn't include those global icons:
![image](https://user-images.githubusercontent.com/1592375/136637214-0876b4d8-7269-4eb8-9852-3e03de44a430.png)

**Comment by Vasmares** [comment]
Aw :/
So also the system app icons cant be used ?
Is there anyway to bypass this ?

**Reply by hufman** [reply]
Indeed, as I understand the car's resource system, only the icon packs that are referenced by the cert are allowed to be used :\

**Comment by Vasmares** [comment]
Hi,

I just checked with my car today.
As stated before, I installed the "old" spotify app again to have the proper sorting.

You said in the last commit, you edited the layout to also be the "global cover art" layout in ID5 even with the old spotify app.

That does not seem to work :/

I have used the latest build from GitHub.
Is this correct ?
Anything else I need to pay attention to ?

**Reply by hufman** [reply]
Yes that should work, I just went out to confirm in my car. Can you peek in the Support tab to see what commit version you are running? It should be one of [these commits](https://github.com/BimmerGestalt/AAIdrive/compare/95a6c57...main).

**Reply by Vasmares** [reply]
Ah ok, so [this ](https://github.com/BimmerGestalt/AAIdrive/releases/)apk does not have the latest commits ?

**Reply by hufman** [reply]
Indeed, any commits added since the last stable release are available in the [Nightly Build](https://bimmergestalt.s3.amazonaws.com/aaidrive/builds/androidautoidrive-latest-main-nomap-sentry-release.apk), available on the main Readme.
